{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "253e263e-15cf-4e5c-8dd4-34631e98b852",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import sqlite3\n",
    "import Coinex_API_Class as Coinex\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "try :\n",
    "    with open('config.json', 'r') as f:\n",
    "        data = json.load(f)\n",
    "        api_key = data['api_key']\n",
    "        api_secret = data['api_secret']\n",
    "        db_name = data['db_name']\n",
    "        num_candles = data['num_candle']\n",
    "        loss_limit = data['loss_limit']\n",
    "        client_id = data['client_id']\n",
    "except Exception as e:\n",
    "    print(f\"Error: {e}\")\n",
    "\n",
    "conn = sqlite3.connect(\"crypto.db\")\n",
    "coinex = Coinex.Coinex_API(api_key, api_secret, connection=conn , client_id=client_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fadbd9c-43b2-4168-9730-9886a799a7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import threading\n",
    "import time\n",
    "\n",
    "def print_numbers():\n",
    "    for i in range(5):\n",
    "        print(f\"عدد {i}\")\n",
    "        time.sleep(1)\n",
    "\n",
    "def print_letters():\n",
    "    for letter in ['A', 'B', 'C', 'D', 'E']:\n",
    "        print(f\"حرف {letter}\")\n",
    "        time.sleep(1.5)\n",
    "\n",
    "# ایجاد دو Thread برای اجرای وظایف مختلف\n",
    "thread1 = threading.Thread(target=print_numbers)\n",
    "thread2 = threading.Thread(target=print_letters)\n",
    "\n",
    "# شروع اجرای Threads\n",
    "thread1.start()\n",
    "thread2.start()\n",
    "\n",
    "# منتظر ماندن تا اتمام اجرای Threads\n",
    "thread1.join()\n",
    "thread2.join()\n",
    "\n",
    "print(\"پایان اجرای همروندی\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f569b06-ba00-4e9f-b4d2-5bed3307f8a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'close': '68169.99',\n",
       "  'high': '68451.65',\n",
       "  'last': '68169.99',\n",
       "  'low': '68035.58',\n",
       "  'market': 'BTCUSDT',\n",
       "  'open': '68445.77',\n",
       "  'period': 86400,\n",
       "  'value': '21572344.9779615622',\n",
       "  'volume': '316.14646532',\n",
       "  'volume_buy': '2.74274679',\n",
       "  'volume_sell': '5.57223814'}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coinex.get_spot_price_ticker(\"BTCUSDT\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19bcfd44-063e-4927-9f0d-2566fd08e674",
   "metadata": {},
   "outputs": [],
   "source": [
    "cursor = conn.cursor()\n",
    "query = \"SELECT * FROM transactions\"  \n",
    "df = pd.read_sql_query(query, conn)\n",
    "df_b = df[df['sold']== 'USDT'].copy()\n",
    "df2=pd.DataFrame()\n",
    "i=0\n",
    "for index, row in df_b.iterrows():\n",
    "    df2.loc[i,'Time_buy']= row['ltime']\n",
    "    df2.loc[i,'ccy'] = row['buy']\n",
    "    df2.loc[i,'pure_amount']= row['balance']\n",
    "    df2.loc[i,'pay_USDT']= row['pay']\n",
    "    df2.loc[i,'Time_sold']= None\n",
    "    df2.loc[i,'recieve']= None\n",
    "    df2.loc[i,'proft']= None \n",
    "    i +=1\n",
    "\n",
    "df2.to_sql('profit', conn, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27228b9b-69c2-4461-a230-b65123c44d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "\n",
    "import asyncio\n",
    "import websockets\n",
    "import json\n",
    "import time\n",
    "import hashlib\n",
    "import gzip\n",
    "import hmac\n",
    "\n",
    "WS_URL = \"wss://socket.coinex.com/v2/spot\"  # Change \"spot\" to \"futures\" when interacting with WS ports\n",
    "access_id = \"47D7C4B286224298BB3D88A9D7161A45\"  # Replace with your access id\n",
    "secret_key = \"F0F2D4156F6892ED2F03576FEE74CFA658C5F8271DCE102D\"  # Replace with your secret key\n",
    "\n",
    "\n",
    "async def ping(conn):\n",
    "    param = {\"method\": \"server.ping\", \"params\": {}, \"id\": 1}\n",
    "    while True:\n",
    "        await conn.send(json.dumps(param))\n",
    "        await asyncio.sleep(3)\n",
    "\n",
    "\n",
    "async def auth(conn):\n",
    "    timestamp = int(time.time() * 1000)\n",
    "\n",
    "    # Generate your signature string\n",
    "    prepared_str = f\"{timestamp}\"\n",
    "    signed_str = (\n",
    "        hmac.new(\n",
    "            bytes(secret_key, \"latin-1\"),\n",
    "            msg=bytes(prepared_str, \"latin-1\"),\n",
    "            digestmod=hashlib.sha256,\n",
    "        )\n",
    "        .hexdigest()\n",
    "        .lower()\n",
    "    )\n",
    "\n",
    "    param = {\n",
    "        \"method\": \"server.sign\",\n",
    "        \"params\": {\n",
    "            \"access_id\": access_id,\n",
    "            \"signed_str\": signed_str,\n",
    "            \"timestamp\": timestamp,\n",
    "        },\n",
    "        \"id\": 1,\n",
    "    }\n",
    "    await conn.send(json.dumps(param))\n",
    "    res = await conn.recv()\n",
    "    res = gzip.decompress(res)\n",
    "    print(\"Authentication Result: \", json.loads(res))\n",
    "\n",
    "\n",
    "async def subscribe_depth(conn):\n",
    "    param = {\n",
    "        \"method\": \"depth.subscribe\",\n",
    "        \"params\": {\"market_list\": [[\"BTCUSDT\", 5, \"0\", True]]},\n",
    "        \"id\": 1,\n",
    "    }\n",
    "    await conn.send(json.dumps(param))\n",
    "    res = await conn.recv()\n",
    "    res = gzip.decompress(res)\n",
    "    print(json.loads(res))\n",
    "\n",
    "\n",
    "async def subscribe_asset(conn):\n",
    "    param = {\"method\": \"balance.subscribe\", \"params\": {\"ccy_list\": [\"USDT\"]}, \"id\": 1}\n",
    "    await conn.send(json.dumps(param))\n",
    "    res = await conn.recv()\n",
    "    res = gzip.decompress(res)\n",
    "    print(json.loads(res))\n",
    "\n",
    "\n",
    "async def main():\n",
    "    try:\n",
    "        # Note: Must close websockets ping feature before creating a new ping task, set ping_interval to None\n",
    "        async with websockets.connect(\n",
    "            uri=WS_URL, compression=None, ping_interval=None\n",
    "        ) as conn:\n",
    "            await auth(conn)\n",
    "            await subscribe_depth(conn)\n",
    "            await subscribe_asset(conn)\n",
    "\n",
    "            asyncio.create_task(ping(conn))\n",
    "\n",
    "            while True:\n",
    "                res = await conn.recv()\n",
    "                res = gzip.decompress(res)\n",
    "                res = json.loads(res)\n",
    "                print(res)\n",
    "    except Exception as e:\n",
    "        print(f\"An error occurred: {e}\")\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    asyncio.run(main())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0c58ee66-2f8c-483c-bab3-017c5f74c7f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.1.1 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\traitlets\\config\\application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\asyncio\\base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\asyncio\\base_events.py\", line 1986, in _run_once\n",
      "    handle._run()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\asyncio\\events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Temp\\ipykernel_12792\\2119920293.py\", line 1, in <module>\n",
      "    import MetaTrader5 as mt5\n",
      "  File \"C:\\Users\\ahads\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\MetaTrader5\\__init__.py\", line 258, in <module>\n",
      "    from ._core import *\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "\nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.1.1 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\numpy\\core\\_multiarray_umath.py:44\u001b[0m, in \u001b[0;36m__getattr__\u001b[1;34m(attr_name)\u001b[0m\n\u001b[0;32m     39\u001b[0m     \u001b[38;5;66;03m# Also print the message (with traceback).  This is because old versions\u001b[39;00m\n\u001b[0;32m     40\u001b[0m     \u001b[38;5;66;03m# of NumPy unfortunately set up the import to replace (and hide) the\u001b[39;00m\n\u001b[0;32m     41\u001b[0m     \u001b[38;5;66;03m# error.  The traceback shouldn't be needed, but e.g. pytest plugins\u001b[39;00m\n\u001b[0;32m     42\u001b[0m     \u001b[38;5;66;03m# seem to swallow it and we should be failing anyway...\u001b[39;00m\n\u001b[0;32m     43\u001b[0m     sys\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mwrite(msg \u001b[38;5;241m+\u001b[39m tb_msg)\n\u001b[1;32m---> 44\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(msg)\n\u001b[0;32m     46\u001b[0m ret \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(_multiarray_umath, attr_name, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m     47\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ret \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[1;31mImportError\u001b[0m: \nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.1.1 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "numpy.core.multiarray failed to import",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mMetaTrader5\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mmt5\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;66;03m# display data on the MetaTrader 5 package\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMetaTrader5 package author: \u001b[39m\u001b[38;5;124m\"\u001b[39m,mt5\u001b[38;5;241m.\u001b[39m__author__)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python312\\Lib\\site-packages\\MetaTrader5\\__init__.py:258\u001b[0m\n\u001b[0;32m    255\u001b[0m RES_E_INTERNAL_FAIL_TIMEOUT         \u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m10005\u001b[39m      \u001b[38;5;66;03m# internal timeout\u001b[39;00m\n\u001b[0;32m    257\u001b[0m \u001b[38;5;66;03m# import C methods to our module\u001b[39;00m\n\u001b[1;32m--> 258\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_core\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;241m*\u001b[39m\n\u001b[0;32m    260\u001b[0m \u001b[38;5;66;03m# internal order send\u001b[39;00m\n\u001b[0;32m    261\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_RawOrder\u001b[39m(order_type, symbol, volume, price, comment\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, ticket\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n",
      "\u001b[1;31mImportError\u001b[0m: numpy.core.multiarray failed to import"
     ]
    }
   ],
   "source": [
    "import MetaTrader5 as mt5\n",
    "# display data on the MetaTrader 5 package\n",
    "print(\"MetaTrader5 package author: \",mt5.__author__)\n",
    "print(\"MetaTrader5 package version: \",mt5.__version__)\n",
    " \n",
    "# establish connection to the MetaTrader 5 terminal\n",
    "if not mt5.initialize():\n",
    "    print(\"initialize() failed, error code =\",mt5.last_error())\n",
    "    quit()\n",
    " \n",
    "# get the number of financial instruments\n",
    "symbols=mt5.symbols_total()\n",
    "if symbols>0:\n",
    "    print(\"Total symbols =\",symbols)\n",
    "else:\n",
    "    print(\"symbols not found\")\n",
    " \n",
    "# shut down connection to the MetaTrader 5 terminal\n",
    "mt5.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e488e452-3dc3-463b-b301-60b5c08d44a4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
